OCPBUGS-36365: LPAR updates to ABI docs (#79373)
openshift/openshift-docs: modules/adding-ibmz-lpar-agent.adoc

// Module included in the following assemblies:
//
// * installing/installing_with_agent_based_installer/prepare-pxe-infra-agent.adoc
// * installing/installing_with_agent_based_installer/installing-with-agent-based-installer.adoc

:_mod-docs-content-type: PROCEDURE
[id="adding-ibmz-lpar-agents_{context}"]

> = Adding {ibm-z-name} Logical Partition (LPAR) as agents

Reviewer: {ibm-z-name} not getting populated in preview

> With {ocp-short} 4.16, you can add the Logical Partition (LPAR) on IBM Z or IBM LinuxONE as a compute node.

Reviewer 2: ```suggestion
Use the following procedure to manually add {ibm-z-name} agents  running in an LPAR. Only use this procedure for {ibm-z-name} clusters running in an LPAR.
```

> 

Reviewer 2: ```suggestion
.Procedure

```

<3> For installations on DASD-type disks, use `rd.dasd` to specify the DASD where Red Hat Enterprise Linux CoreOS (RHCOS) is to be installed. For installations on FCP-type disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where RHCOS is to be installed.
<4> Specify this parameter when you use an Open Systems Adapter (OSA) or HiperSockets.

. Generate the `.ins` and `initrd.img.addrsize` files.
+
The `.ins` file is a special file that includes installation data and is present on the FTP server. It can be accessed from the HMC system.
+
This file contains details such as mapping of the location of installation data on the disk or FTP server, the memory locations where the data is to be copied.
+
*Note*: In {ocp-short} 4.16, the `.ins` file and 'initrd.img.addrsize' are not automatically generated as part of boot-artifacts from the installer. You must manually generate these files.

.. Run the following commands to get the size of the `kernel` and `initrd`: 
+
[source,yaml]
----
KERNEL_IMG_PATH='./kernel.img'
INITRD_IMG_PATH='./initrd.img'
CMDLINE_PATH='./generic.prm'
kernel_size=$(stat -c%s $KERNEL_IMG_PATH )
initrd_size=$(stat -c%s $INITRD_IMG_PATH)
----

> .. Round the `kernel` size up to the next MiB boundary. This value  is the starting address of `initrd.img`.

Reviewer 2: Is MiB correct here? Or should it rather be MB? 
```suggestion
.. Round the `kernel` size up to the next MB boundary. This value is the starting address of `initrd.img`.
```

+

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```
openshift-docs guidelines are different than ACM. source,bash is only used for bash scripts see: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

----
offset=$(( (kernel_size + 1048575) / 1048576 * 1048576 ))
----
.. Create the kernel binary patch file that contains the `initrd` address and size by running the following commands:
+

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

----
INITRD_IMG_NAME=$(echo $INITRD_IMG_PATH | rev | cut -d '/' -f 1 | rev)
KERNEL_OFFSET=0x00000000
KERNEL_CMDLINE_OFFSET=0x00010480
INITRD_ADDR_SIZE_OFFSET=0x00010408
OFFSET_HEX=$(printf '0x%08x\n' $offset)
----

.. Convert the address and size to binary format by running the following commands:
+

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

----
printf "$(printf '%016x\n' $initrd_size)" | xxd -r -p > temp_size.bin
----

.. Concatenate the address and size binaries by running the following command:
+

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

----
cat temp_address.bin temp_size.bin > "$INITRD_IMG_NAME.addrsize"
----

.. Clean up temporary files by running the following command:
+

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

// Module included in the following assemblies:
//
// * installing/installing_with_agent_based_installer/prepare-pxe-infra-agent.adoc

:_mod-docs-content-type: PROCEDURE
[id="adding-ibmz-lpar-agents_{context}"]

> = Adding {ibm-z-title} Logical Partition (LPAR) as agents

Reviewer: How is this getting populated? {ibm-z-title}

Author: <img width="782" alt="image" src="https://github.com/user-attachments/assets/e045173e-ff64-471b-87a4-f8657d26d9f0">
Like this

> Use the following procedure to manually add {ibm-z-name} agents running in an LPAR. Use this procedure only for {ibm-z-name} clusters running in an LPAR.

Reviewer 3: ```suggestion
Use the following procedure to manually add {ibm-z-name} agents to your cluster that runs in an LPAR environment. Use this procedure only for {ibm-z-name} clusters running in an LPAR.
```

.Procedure

. Create a boot parameter file for the agents.
+
.Sample parameter file:
[source,yaml]
----

> rd.neednet=1 cio_ignore=all,!condev \

Reviewer 3: Just checking if this is a YAML file or another type of file? The file structure here differs from our conventional YAML file.

Reviewer 2: I agree this is a parm file I suggest to use` [source,terminal]`

console=ttysclp0 \
ignition.firstboot ignition.platform.id=metal
coreos.live.rootfs_url=http://<http_server>/rhcos-<version>-live-rootfs.<architecture>.img \// <1>
coreos.inst.persistent-kargs=console=ttysclp0
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> \// <2>
rd.znet=qeth,<network_adaptor_range>,layer2=1
rd.<disk_type>=<adapter> \// <3>
zfcp.allow_lun_scan=0
ai.ip_cfg_override=1 \// <4>
random.trust_cpu=on rd.luks.options=discard
----
<1> For the `coreos.live.rootfs_url` artifact, specify the matching `rootfs` artifact for the `kernel` and `initramfs` that you are starting. Only HTTP and HTTPS protocols are supported.
<2> For the `ip` parameter, manually assign the IP address, as described in _Installing a cluster with z/VM on IBM Z and IBM LinuxONE_. 

> <3> For installations on DASD-type disks, use `rd.dasd` to specify the DASD where Red Hat Enterprise Linux CoreOS (RHCOS) is to be installed. For installations on FCP-type disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where RHCOS is to be installed.

Reviewer 3: Consider spelling out DASD as I do not see it's full name in the overall document.

Reviewer 2: I would leave it as is. DASD is a common term for IBM Z customers and we spelled it out once in the UPI docs in the minimum requirements.

<4> Specify this parameter when you use an Open Systems Adapter (OSA) or HiperSockets.

. Generate the `.ins` and `initrd.img.addrsize` files.
+

> The `.ins` file is a special file that includes installation data and is present on the FTP server. It can be accessed from the HMC system.

Reviewer 3: ```suggestion
The `.ins` file is a special file that includes installation data and is present on the FTP server. The file can be accessed from the HMC system.
```

Reviewer 3: Spell out HCM?

Reviewer 2: Same as before we already have that IBM Z term spelled out in the UPI installation.

+

> This file contains details such as mapping of the location of installation data on the disk or FTP server, the memory locations where the data is to be copied.

Reviewer 3: Add to the previous sentence to form a block paragraph.

+

> *Note*: The `.ins` file and `initrd.img.addrsize` are not automatically generated as part of boot-artifacts from the installer. You must manually generate these files.

Reviewer 3: Best to follow the typical admonition block for a note:

```
[NOTE]
====

====
```

> *Note*: The `.ins` file and `initrd.img.addrsize` are not automatically generated as part of boot-artifacts from the installer. You must manually generate these files.

Reviewer 3: installer = installation program?

Author: yes. Essentially, they both mean the same thing, right?

.. Run the following commands to get the size of the `kernel` and `initrd`: 
+
[source,terminal]
----
KERNEL_IMG_PATH='./kernel.img'
INITRD_IMG_PATH='./initrd.img'
CMDLINE_PATH='./generic.prm'
kernel_size=$(stat -c%s $KERNEL_IMG_PATH )
initrd_size=$(stat -c%s $INITRD_IMG_PATH)
----

> .. Round the `kernel` size up to the next MB boundary. This value  is the starting address of `initrd.img`.

Reviewer 3: ```suggestion
.. Round the `kernel` size up to the next MB boundary. This value  is the starting address of `initrd.img`:
```

I assume a customer must add the line to their file? Maybe call this out?

Author: The customer does not need to add this line. It is part of the script. We are only giving details on how to identify the value.So, no doc changes required.

> .. Round the `kernel` size up to the next MB boundary. This value  is the starting address of `initrd.img`.

Reviewer 3: Silly nit but maybe "megabytes (MB)"  would be clearer as no value prepends MB.

+
[source,terminal]
----
offset=$(( (kernel_size + 1048575) / 1048576 * 1048576 ))
----

> .. Create the kernel binary patch file that contains the `initrd` address and size by running the following commands:

Reviewer 3: Our [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#single-command-per-code-block) state to add a single command per code block. I'm not sure if the commands need to be run in a CLI or added to a file? If added to a file and the file is run as part of a script then one command per code block would not apply.

Author: Changed the sentence to say "....running the following script:"

+
[source,terminal]
----
INITRD_IMG_NAME=$(echo $INITRD_IMG_PATH | rev | cut -d '/' -f 1 | rev)
KERNEL_OFFSET=0x00000000
KERNEL_CMDLINE_OFFSET=0x00010480
INITRD_ADDR_SIZE_OFFSET=0x00010408
OFFSET_HEX=$(printf '0x%08x\n' $offset)
----

> .. Convert the address and size to binary format by running the following commands:

Reviewer 3: ```suggestion
.. Convert the address and size to binary format by running the following command:
```

+
[source,terminal]
----
printf "$(printf '%016x\n' $initrd_size)" | xxd -r -p > temp_size.bin
----

.. Concatenate the address and size binaries by running the following command:
+
[source,terminal]
----

> cat temp_address.bin temp_size.bin > "$INITRD_IMG_NAME.addrsize"

Reviewer 3: None of these commands start with a $ sign so I assume they are added to a file and run together in a script format?

Author: It is a python script, so i will leave it as is.

----

> .. Clean up temporary files by running the following command:

Reviewer 3: ```suggestion
.. Delete temporary files by running the following command:
```

+
[source,terminal]
----
rm -rf temp_address.bin temp_size.bin
----

.. Create the `.ins` file. The file is based on the paths of the `kernel.img`, `initrd.img`, `initrd.img.addrsize`, and `cmdline` files and the memory locations where the data is to be copied.
+
[source,terminal]
----
$KERNEL_IMG_PATH $KERNEL_OFFSET
$INITRD_IMG_PATH $OFFSET_HEX
$INITRD_IMG_NAME.addrsize $INITRD_ADDR_SIZE_OFFSET
$CMDLINE_PATH $KERNEL_CMDLINE_OFFSET
----

> . Transfer the `initrd`, `kernel`, `generic.ins`, and `initrd.img.addrsize` parameter files to the file server. For more information about how to transfer the files with FTP and boot, see link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_a_standard_rhel_8_installation/installing-in-an-lpar_installing-rhel[Installing in an LPAR].    

Reviewer 3: This link brings me to the RHEL 8 homepage. Was this intended?

![image](https://github.com/user-attachments/assets/94e7de78-99bf-4837-86aa-d549df97af35)

Reviewer 2: Seems like the RHEL docs have been reorganized we also have that link in the UPI docs it worked a few months ago. I'll check where we can link to.

Reviewer 2: @krishnanshruthi Seems like the entire IBM Z documentation has vanished from the RHEL docs. Please take this link instead: https://www.ibm.com/docs/en/linux-on-systems?topic=bl-booting-linux-in-lpar-mode

Reviewer 2: @alishaIBM I don't see IBM Power in the RHEL docs anymore either. Just in case you also have links to RHEL

Reviewer 3: The link still points to the RHEL CP homepage but this is OK?

// Module included in the following assemblies:
//
// * installing/installing_with_agent_based_installer/prepare-pxe-infra-agent.adoc

:_mod-docs-content-type: PROCEDURE
[id="adding-ibmz-lpar-agents_{context}"]
= Adding {ibm-z-title} Logical Partition (LPAR) as agents

Use the following procedure to manually add {ibm-z-name} agents to your cluster that runs in an LPAR environment. Use this procedure only for {ibm-z-name} clusters running in an LPAR.

> .Procedure

Reviewer 4: Since this procedure involves using python, I think some prereq about having it installed would be important to mention here:
```suggestion
.Prerequisites

* You have Python 3 installed.

.Procedure
```

----
rd.neednet=1 cio_ignore=all,!condev \
console=ttysclp0 \
ignition.firstboot ignition.platform.id=metal
coreos.live.rootfs_url=http://<http_server>/rhcos-<version>-live-rootfs.<architecture>.img \// <1>
coreos.inst.persistent-kargs=console=ttysclp0
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> \// <2>
rd.znet=qeth,<network_adaptor_range>,layer2=1
rd.<disk_type>=<adapter> \// <3>
zfcp.allow_lun_scan=0
ai.ip_cfg_override=1 \// <4>
random.trust_cpu=on rd.luks.options=discard
----
<1> For the `coreos.live.rootfs_url` artifact, specify the matching `rootfs` artifact for the `kernel` and `initramfs` that you are starting. Only HTTP and HTTPS protocols are supported.
<2> For the `ip` parameter, manually assign the IP address, as described in _Installing a cluster with z/VM on IBM Z and IBM LinuxONE_. 
<3> For installations on DASD-type disks, use `rd.dasd` to specify the DASD where Red Hat Enterprise Linux CoreOS (RHCOS) is to be installed. For installations on FCP-type disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where RHCOS is to be installed.
<4> Specify this parameter when you use an Open Systems Adapter (OSA) or HiperSockets.

> . Generate the `.ins` and `initrd.img.addrsize` files.

Reviewer 4: I believe that this step needs to be rewritten in a way that's simpler for users to understand. It seems like this is really one large script, where all these commands can be run in a single batch. 

If that's the case, then I understand the attempt to describe what each group of commands is doing in a substep, but I believe that creates this weird situation where it looks like we're telling users to run a series of commands in a command line, but really we're expecting them to run everything at once. This creates an issue both in terms of style guide rules as well as the logical flow of the procedure.

So if all the commands in this step need to be run in a single script, I think the following changes need to be done:

1.  The commands in all of the substeps, 2a-2g, should be placed in one large code block so that everything is together, and it is made clear that this all needs to be run as a single script
2. If you still want to describe what the individual commands in the script are doing, then find some way to add that description. I think the most straightforward method would be to use callout labels like you do in step 1 of this procedure. There are other ways, such as including a paragraph before the code block to explain what is happening, but I think the callouts would look the best.

Reviewer 4: Also, it would be best to explicitly mention in the procedure that users need to run a python script specifically:
```suggestion
. Generate the `.ins` and `initrd.img.addrsize` files by running the following Python script:
+
[source,terminal]
```

+
The `.ins` file is a special file that includes installation data and is present on the FTP server. It can be accessed from the HMC system.
This file contains details such as mapping of the location of installation data on the disk or FTP server, the memory locations where the data is to be copied.

[NOTE]
====
The `.ins` file and `initrd.img.addrsize` are not automatically generated as part of boot-artifacts from the installer. You must manually generate these files.
====

.. Run the following commands to get the size of the `kernel` and `initrd`: 
+
[source,terminal]
----
KERNEL_IMG_PATH='./kernel.img'
INITRD_IMG_PATH='./initrd.img'
CMDLINE_PATH='./generic.prm'

> kernel_size=$(stat -c%s $KERNEL_IMG_PATH )

Reviewer 3: ```suggestion
kernel_size=$(stat -c%s $KERNEL_IMG_PATH)
```

Author: The values for line 52 is auto-populated by the command on line 49. So, no doc changes required.

initrd_size=$(stat -c%s $INITRD_IMG_PATH)
----

> .. Round the `kernel` size up to the next megabytes (MB) boundary. This value is the starting address of `initrd.img`: 

Reviewer 4: Since the second sentence doesn't mention anything about the command that follows,  I think you can just remove the colon and use a period instead:
```suggestion
.. Round the `kernel` size up to the next megabytes (MB) boundary. This value is the starting address of `initrd.img`.
```

// Module included in the following assemblies:
//
// * installing/installing_with_agent_based_installer/prepare-pxe-infra-agent.adoc

:_mod-docs-content-type: PROCEDURE
[id="adding-ibmz-lpar-agents_{context}"]
= Adding {ibm-z-title} agents in a Logical Partition (LPAR)

Use the following procedure to manually add {ibm-z-name} agents to your cluster that runs in an LPAR environment. Use this procedure only for {ibm-z-name} clusters running in an LPAR.

.Prerequisites

> * You have Python 3 installed.

Reviewer 3: Rendering issue.

![Screenshot from 2024-08-14 10-10-57](https://github.com/user-attachments/assets/1668652d-00e3-4d30-9475-adda52ecd1a5)

Reviewer 4: +1, looks like you need an extra blank like here:
```suggestion
* You have Python 3 installed.

```

.Procedure

. Create a boot parameter file for the agents.
+
.Example parameter file
[source,terminal]
----
rd.neednet=1 cio_ignore=all,!condev \
console=ttysclp0 \
ignition.firstboot ignition.platform.id=metal
coreos.live.rootfs_url=http://<http_server>/rhcos-<version>-live-rootfs.<architecture>.img \// <1>
coreos.inst.persistent-kargs=console=ttysclp0
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> \// <2>
rd.znet=qeth,<network_adaptor_range>,layer2=1
rd.<disk_type>=<adapter> \// <3>
zfcp.allow_lun_scan=0
ai.ip_cfg_override=1 \// <4>
random.trust_cpu=on rd.luks.options=discard
----
<1> For the `coreos.live.rootfs_url` artifact, specify the matching `rootfs` artifact for the `kernel` and `initramfs` that you are starting. Only HTTP and HTTPS protocols are supported.
<2> For the `ip` parameter, manually assign the IP address, as described in _Installing a cluster with z/VM on IBM Z and IBM LinuxONE_.

> <3> For installations on DASD-type disks, use `rd.dasd` to specify the DASD where RedHat Enterprise Linux CoreOS (RHCOS) is to be installed. For installations on FCP-type disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where RHCOS is to be installed.

Reviewer 3: common-attributes.adoc contains an attribute for RedHat Enterprise Linux CoreOS (RHCOS). I would recommend to use it in the event of a name change. One never knows! Same for "RHCOS".

<4> Specify this parameter when you use an Open Systems Adapter (OSA) or HiperSockets.

. Generate the `.ins` and `initrd.img.addrsize` files.
+
The `.ins` file is a special file that includes installation data and is present on the FTP server. It can be accessed from the HMC system.
This file contains details such as mapping of the location of installation data on the disk or FTP server, the memory locations where the data is to be copied.
+
[NOTE]
====
The `.ins` and `initrd.img.addrsize` files are not automatically generated as part of boot-artifacts from the installer. You must manually generate these files.
====

.. Save the following script to a file, such as `generate-files.py`:
+

> .`generate-files.py` file

Reviewer 3: ```suggestion
.`Example of a Python file named `generate-files.py`
```

Reviewer 4: If you do use this suggestion, there seems to be an extra backtick at the start of that suggestion:
```suggestion
.Example of a Python file named `generate-files.py` file
```

[source,python]
----

> # The following commands retrieve the size of the `kernel` and `initrd`:

Reviewer 3: I don't think the dollar sign means much in Python code when it comes to running Python commands. Worth checking with the SME. With functions, methods, and so on that are called rather than run, the dollar sign would only confuse the naming convention for Python instances.

> # The following commands retrieve the size of the `kernel` and `initrd`:

Reviewer 3: I see duplicated includes?

![Screenshot from 2024-08-14 10-46-11](https://github.com/user-attachments/assets/0051111c-9949-4345-9605-3e39985a3ec2)

offset=$(( (kernel_size + 1048575) / 1048576 * 1048576 ))
INITRD_IMG_NAME=$(echo $INITRD_IMG_PATH | rev | cut -d '/' -f 1 | rev)
# The following commands create the kernel binary patch file that contains the `initrd` address and size:
KERNEL_OFFSET=0x00000000
KERNEL_CMDLINE_OFFSET=0x00010480
INITRD_ADDR_SIZE_OFFSET=0x00010408
OFFSET_HEX=$(printf '0x%08x\n' $offset)
# The following command converts the address and size to binary format:
printf "$(printf '%016x\n' $initrd_size)" | xxd -r -p > temp_size.bin
# The following command concatenates the address and size binaries:
cat temp_address.bin temp_size.bin > "$INITRD_IMG_NAME.addrsize"
# The following command deletes temporary files:
rm -rf temp_address.bin temp_size.bin
# The following commands create the `.ins` file.
# The file is based on the paths of the `kernel.img`, `initrd.img`, `initrd.img.addrsize`, and `cmdline` files and the memory locations where the data is to be copied.
$KERNEL_IMG_PATH $KERNEL_OFFSET
$INITRD_IMG_PATH $OFFSET_HEX
$INITRD_IMG_NAME.addrsize $INITRD_ADDR_SIZE_OFFSET
$CMDLINE_PATH $KERNEL_CMDLINE_OFFSET
----

.. Execute the script by running the following command:
+
[source,terminal]
----

> $ python3 generate-file.py

Reviewer 3: ```suggestion
$ python3 <file_name>.py
```

----

> . Transfer the `initrd`, `kernel`, `generic.ins`, and `initrd.img.addrsize` parameter files to the file server. For more information, see link:https://www.ibm.com/docs/en/linux-on-systems?topic=bl-booting-linux-in-lpar-mode[Booting Linux in LPAR mode].

Reviewer 3: ```suggestion
. Transfer the `initrd`, `kernel`, `generic.ins`, and `initrd.img.addrsize` parameter files to the file server. For more information, see link:https://www.ibm.com/docs/en/linux-on-systems?topic=bl-booting-linux-in-lpar-mode[Booting Linux in LPAR mode (IBM documentation].
```

This aspect is importent to let a customer know that they will need to leave our customer documentation to visit IBM's documentation portal, which luckily does not currently require a login operation.
